src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs

```diff
@@ -50,5 +50,23 @@ public byte[] WriteMessage(HubMessage hubMessage)
                 return _dataEncoder.Encode(ms.ToArray());
             }
         }
+
+        public override bool Equals(object obj)
+        {
+            var readerWriter = obj as HubProtocolReaderWriter;
+            if (readerWriter == null)
+            {
+                return false;
+            }
+
+            // Note: ReferenceEquals on HubProtocol works for our implementation of IHubProtocolResolver because we use Singletons from DI
+            // However if someone replaces the implementation and returns a new ProtocolResolver for every connection they wont get the perf benefits
```

Comment:
TBH, still nervous about this. It's not even that they won't get the perf benefits, they'll get significantly more memory usage (and `O(n)` memory growth where `n` is the number of connections).

Reply:
Also, the client is in control of how much this memory grows.

Reply:
How is it client controlled? The server is the one that configures the amount of protocols supported? Are you taking about in theory you can return a different instance of a protocol per connection? Again how is that a real concern?